File: wizards/com/sun/star/wizards/common/NumberFormatter.py
#
# This file is part of the LibreOffice project.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# This file incorporates work covered by the following license notice:
#
#   Licensed to the Apache Software Foundation (ASF) under one or more
#   contributor license agreements. See the NOTICE file distributed
#   with this work for additional information regarding copyright
#   ownership. The ASF licenses this file to you under the Apache
#   License, Version 2.0 (the "License"); you may not use this file
#   except in compliance with the License. You may obtain a copy of
#   the License at http://www.apache.org/licenses/LICENSE-2.0 .
#
import traceback

from com.sun.star.lang import Locale
from com.sun.star.util.NumberFormat import DATE, LOGICAL, DATETIME, TEXT, NUMBER

class NumberFormatter(object):

    def __init__(self, _xNumberFormatsSupplier, _aLocale, _xMSF=None):
        self.iDateFormatKey = -1
        self.iDateTimeFormatKey = -1
        self.iNumberFormatKey = -1
        self.iTextFormatKey = -1
        self.iTimeFormatKey = -1
        self.iLogicalFormatKey = -1
        self.bNullDateCorrectionIsDefined = False
        self.aLocale = _aLocale
        if _xMSF is not None:
            self.xNumberFormatter = _xMSF.createInstance(
                "com.sun.star.util.NumberFormatter")
        self.xNumberFormats = _xNumberFormatsSupplier.NumberFormats
        self.xNumberFormatSettings = \
            _xNumberFormatsSupplier.NumberFormatSettings
        self.xNumberFormatter.attachNumberFormatsSupplier(
            _xNumberFormatsSupplier)

    '''
    @param _xMSF
    @param _xNumberFormatsSupplier
    @return
    @throws Exception
    @deprecated
    '''

    @classmethod
    def createNumberFormatter(self, _xMSF, _xNumberFormatsSupplier):
        oNumberFormatter = _xMSF.createInstance(
            "com.sun.star.util.NumberFormatter")
        oNumberFormatter.attachNumberFormatsSupplier(_xNumberFormatsSupplier)
        return oNumberFormatter

    '''
    gives a key to pass to a NumberFormat object. <br/>
    example: <br/>
    <pre>
    XNumberFormatsSupplier nsf =
        (XNumberFormatsSupplier)UnoRuntime.queryInterface(...,document)
    int key = Desktop.getNumberFormatterKey(
        nsf, ...star.i18n.NumberFormatIndex.DATE...)
    XNumberFormatter nf = Desktop.createNumberFormatter(xmsf, nsf);
    nf.convertNumberToString( key, 1972 );
    </pre>
    @param numberFormatsSupplier
    @param type - a constant out of i18n.NumberFormatIndex enumeration.
    @return a key to use with a util.NumberFormat instance.
    '''

    @classmethod
    def getNumberFormatterKey(self, numberFormatsSupplier, Type):
        return numberFormatsSupplier.NumberFormats.getFormatIndex(
            Type, Locale())

    def convertNumberToString(self, _nkey, _dblValue, _xNumberFormatter=None):
        if _xNumberFormatter is None:
            return self.xNumberFormatter.convertNumberToString(
                _nkey, _dblValue)
        else:
            return _xNumberFormatter.convertNumberToString(_nkey, _dblValue)

    def convertStringToNumber(self, _nkey, _sString):
        return self.xNumberFormatter.convertStringToNumber(_nkey, _sString)

    '''
    @param dateCorrection The lDateCorrection to set.
    '''

    def setNullDateCorrection(self, dateCorrection):
        self.lDateCorrection = dateCorrection

    def defineNumberFormat(self, _FormatString):
        try:
            NewFormatKey = self.xNumberFormats.queryKey(
                _FormatString, self.aLocale, True)
            if NewFormatKey is -1:
                NewFormatKey = self.xNumberFormats.addNew(
                    _FormatString, self.aLocale)

            return NewFormatKey
        except Exception, e:
            traceback.print_exc()
            return -1

    '''
    returns a numberformat for a FormatString.
    @param _FormatString
    @param _aLocale
    @return
    '''

    def defineNumberFormat(self, _FormatString, _aLocale):
        try:
            NewFormatKey = self.xNumberFormats.queryKey(
                _FormatString, _aLocale, True)
            if NewFormatKey == -1:
                NewFormatKey = self.xNumberFormats.addNew(
                    _FormatString, _aLocale)

            return NewFormatKey
        except Exception, e:
            traceback.print_exc()
            return -1

    def setNumberFormat(self, _xFormatObject, _FormatKey, _oNumberFormatter):
        try:
            xNumberFormat = _oNumberFormatter.xNumberFormats.getByKey(
                _FormatKey)
            FormatString = str(Helper.getUnoPropertyValue(
                xNumberFormat, "FormatString"))
            oLocale = Helper.getUnoPropertyValue(xNumberFormat, "Locale")
            NewFormatKey = defineNumberFormat(FormatString, oLocale)
            _xFormatObject.setPropertyValue(
                "FormatsSupplier",
                _oNumberFormatter.xNumberFormatter.getNumberFormatsSupplier())
            if _xFormatObject.getPropertySetInfo().hasPropertyByName(
                    "NumberFormat"):
                _xFormatObject.setPropertyValue("NumberFormat", NewFormatKey)
            elif _xFormatObject.getPropertySetInfo().hasPropertyByName(
                    "FormatKey"):
                _xFormatObject.setPropertyValue("FormatKey", NewFormatKey)
            else:
                # TODO: throws a exception in a try catch environment, very helpful?
                raise Exception

        except Exception, exception:
            traceback.print_exc()

    def getNullDateCorrection(self):
        if not self.bNullDateCorrectionIsDefined:
            dNullDate = Helper.getUnoStructValue(
                self.xNumberFormatSettings, "NullDate")
            lNullDate = Helper.convertUnoDatetoInteger(dNullDate)
            oCal = java.util.Calendar.getInstance()
            oCal.set(1900, 1, 1)
            dTime = oCal.getTime()
            lTime = dTime.getTime()
            lDBNullDate = lTime / (3600 * 24000)
            self.lDateCorrection = lDBNullDate - lNullDate
            return self.lDateCorrection
        else:
            return self.lDateCorrection

    def setBooleanReportDisplayNumberFormat(self):
        FormatString = "[=1]" + str(9745) + ";[=0]" + str(58480) + ";0"
        self.iLogicalFormatKey = self.xNumberFormats.queryKey(
            FormatString, self.aLocale, True)
        try:
            if self.iLogicalFormatKey == -1:
                self.iLogicalFormatKey = self.xNumberFormats.addNew(
                    FormatString, self.aLocale)

        except Exception, e:
            #MalformedNumberFormat
            traceback.print_exc()
            self.iLogicalFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.LOGICAL, self.aLocale)

        return self.iLogicalFormatKey

    '''
    @return Returns the iDateFormatKey.
    '''

    def getDateFormatKey(self):
        if self.iDateFormatKey == -1:
            self.iDateFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.DATE, self.aLocale)

        return self.iDateFormatKey

    '''
    @return Returns the iDateTimeFormatKey.
    '''

    def getDateTimeFormatKey(self):
        if self.iDateTimeFormatKey == -1:
            self.iDateTimeFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.DATETIME, self.aLocale)

        return self.iDateTimeFormatKey

    '''
    @return Returns the iLogicalFormatKey.
    '''

    def getLogicalFormatKey(self):
        if self.iLogicalFormatKey == -1:
            self.iLogicalFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.LOGICAL, self.aLocale)

        return self.iLogicalFormatKey

    '''
    @return Returns the iNumberFormatKey.
    '''

    def getNumberFormatKey(self):
        if self.iNumberFormatKey == -1:
            self.iNumberFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.NUMBER, self.aLocale)

        return self.iNumberFormatKey

    '''
    @return Returns the iTextFormatKey.
    '''

    def getTextFormatKey(self):
        if self.iTextFormatKey == -1:
            self.iTextFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.TEXT, self.aLocale)

        return self.iTextFormatKey

    '''
    @return Returns the iTimeFormatKey.
    '''

    def getTimeFormatKey(self):
        if self.iTimeFormatKey == -1:
            self.iTimeFormatKey = self.xNumberFormats.getStandardFormat(
                NumberFormat.TIME, self.aLocale)

        return self.iTimeFormatKey

File: wizards/com/sun/star/wizards/common/NumberFormatter.java
/*
 * This file is part of the LibreOffice project.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * This file incorporates work covered by the following license notice:
 *
 *   Licensed to the Apache Software Foundation (ASF) under one or more
 *   contributor license agreements. See the NOTICE file distributed
 *   with this work for additional information regarding copyright
 *   ownership. The ASF licenses this file to you under the Apache
 *   License, Version 2.0 (the "License"); you may not use this file
 *   except in compliance with the License. You may obtain a copy of
 *   the License at http://www.apache.org/licenses/LICENSE-2.0 .
 */

package com.sun.star.wizards.common;

import java.util.Date;

import com.sun.star.beans.XPropertySet;
import com.sun.star.lang.Locale;
import com.sun.star.lang.XMultiServiceFactory;
import com.sun.star.uno.AnyConverter;
import com.sun.star.uno.UnoRuntime;
import com.sun.star.uno.XInterface;
import com.sun.star.util.NumberFormat;
import com.sun.star.util.XNumberFormatTypes;
import com.sun.star.util.XNumberFormats;
import com.sun.star.util.XNumberFormatsSupplier;
import com.sun.star.util.XNumberFormatter;


public class NumberFormatter
{

    private int iDateFormatKey = -1;
    private int iDateTimeFormatKey = -1;
    private int iNumberFormatKey = -1;
    private int iTextFormatKey = -1;
    private int iTimeFormatKey = -1;
    private int iLogicalFormatKey = -1;
    private long lDateCorrection;
    private XNumberFormatter xNumberFormatter;
    private XNumberFormats xNumberFormats;
    private XNumberFormatTypes xNumberFormatTypes;
    private XPropertySet xNumberFormatSettings;
    private final Locale aLocale;
    static private final Locale enLocale = new Locale( "en", "US", "" );


    public NumberFormatter(XMultiServiceFactory _xMSF, XNumberFormatsSupplier _xNumberFormatsSupplier, Locale _aLocale) throws Exception
    {
        aLocale = _aLocale;
        Object oNumberFormatter = _xMSF.createInstance("com.sun.star.util.NumberFormatter");
        xNumberFormats = _xNumberFormatsSupplier.getNumberFormats();
        xNumberFormatSettings = _xNumberFormatsSupplier.getNumberFormatSettings();
        xNumberFormatter = UnoRuntime.queryInterface(XNumberFormatter.class, oNumberFormatter);
        xNumberFormatter.attachNumberFormatsSupplier(_xNumberFormatsSupplier);
        xNumberFormatTypes = UnoRuntime.queryInterface(XNumberFormatTypes.class, xNumberFormats);

    }

    public NumberFormatter(XNumberFormatsSupplier _xNumberFormatsSupplier, Locale _aLocale) throws Exception
    {
        aLocale = _aLocale;
        xNumberFormats = _xNumberFormatsSupplier.getNumberFormats();
        xNumberFormatSettings = _xNumberFormatsSupplier.getNumberFormatSettings();
        xNumberFormatTypes = UnoRuntime.queryInterface(XNumberFormatTypes.class, xNumberFormats);
    }


    public double convertStringToNumber(int _nkey, String _sString)throws Exception
    {
        return xNumberFormatter.convertStringToNumber(_nkey, _sString);
    }


    /**
     * @param dateCorrection The lDateCorrection to set.
     */
    public void setNullDateCorrection(long dateCorrection)
    {
        lDateCorrection = dateCorrection;
    }


    public int defineNumberFormat(String _FormatString)
    {
        try
        {
            int NewFormatKey = xNumberFormats.queryKey(_FormatString, aLocale, true);
            if (NewFormatKey == -1)
            {
                NewFormatKey = xNumberFormats.addNewConverted(_FormatString, enLocale, aLocale);
            }
            return NewFormatKey;
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
            return -1;
        }
    }


    /**
     * returns a numberformat for a FormatString.
     */
    private int defineNumberFormat(String _FormatString, Locale _aLocale)
    {
        try
        {
            int NewFormatKey = xNumberFormats.queryKey(_FormatString, _aLocale, true);
            if (NewFormatKey == -1)
            {
                NewFormatKey = xNumberFormats.addNew(_FormatString, _aLocale);
            }
            return NewFormatKey;
        }
        catch (Exception e)
        {
            e.printStackTrace(System.err);
            return -1;
        }
    }



    public void setNumberFormat(XInterface _xFormatObject, int _FormatKey, NumberFormatter _oNumberFormatter)
    {
        try
        {
            XPropertySet xNumberFormat = _oNumberFormatter.xNumberFormats.getByKey(_FormatKey);
            String FormatString = AnyConverter.toString(Helper.getUnoPropertyValue(xNumberFormat, "FormatString"));
            Locale oLocale = (Locale) Helper.getUnoPropertyValue(xNumberFormat, "Locale");
            int NewFormatKey = defineNumberFormat(FormatString, oLocale);
            XPropertySet xPSet = UnoRuntime.queryInterface(XPropertySet.class, _xFormatObject);
            if (xPSet.getPropertySetInfo().hasPropertyByName("FormatsSupplier"))
            {
                xPSet.setPropertyValue("FormatsSupplier", _oNumberFormatter.xNumberFormatter.getNumberFormatsSupplier());
            }
            if (xPSet.getPropertySetInfo().hasPropertyByName("NumberFormat"))
            {
                xPSet.setPropertyValue("NumberFormat", Integer.valueOf(NewFormatKey));
            }
            else if (xPSet.getPropertySetInfo().hasPropertyByName("FormatKey"))
            {
                xPSet.setPropertyValue("FormatKey", Integer.valueOf(NewFormatKey));
            }
            else
            {
                // TODO: throws a exception in a try catch environment, very helpful?
                throw new Exception();
            }
        }
        catch (Exception exception)
        {
            exception.printStackTrace(System.err);
        }
    }


    public long getNullDateCorrection()
    {
        com.sun.star.util.Date dNullDate = (com.sun.star.util.Date) Helper.getUnoStructValue(this.xNumberFormatSettings, "NullDate");
        long lNullDate = Helper.convertUnoDatetoInteger(dNullDate);
        java.util.Calendar oCal = java.util.Calendar.getInstance();
        oCal.set(1900, 1, 1);
        Date dTime = oCal.getTime();
        long lTime = dTime.getTime();
        long lDBNullDate = lTime / (3600 * 24000);
        lDateCorrection = lDBNullDate - lNullDate;
        return lDateCorrection;
    }


    public int setBooleanReportDisplayNumberFormat()
    {
        String FormatString = "[=1]" + '"' + (char)9745 + '"' + ";[=0]" + '"' + (char)58480 + '"' + ";0";
        iLogicalFormatKey = xNumberFormats.queryKey(FormatString, aLocale, true);
        try
        {
            if (iLogicalFormatKey == -1)
            {
                iLogicalFormatKey = xNumberFormats.addNew(FormatString, aLocale);
            }
        }
        catch (Exception e)
        {         //MalformedNumberFormat
            e.printStackTrace();
            iLogicalFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.LOGICAL, aLocale);
        }
        return iLogicalFormatKey;
    }


    /**
     * @return Returns the iDateFormatKey.
     */
    public int getDateFormatKey()
    {
        if (iDateFormatKey == -1)
        {
            iDateFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.DATE, aLocale);
        }
        return iDateFormatKey;
    }
    /**
     * @return Returns the iDateTimeFormatKey.
     */
    public int getDateTimeFormatKey()
    {
        if (iDateTimeFormatKey == -1)
        {
            iDateTimeFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.DATETIME, aLocale);
        }
        return iDateTimeFormatKey;
    }
    /**
     * @return Returns the iLogicalFormatKey.
     */
    public int getLogicalFormatKey()
    {
        if (iLogicalFormatKey == -1)
        {
            iLogicalFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.LOGICAL, aLocale);
        }
        return iLogicalFormatKey;
    }
    /**
     * @return Returns the iNumberFormatKey.
     */
    public int getNumberFormatKey()
    {
        if (iNumberFormatKey == -1)
        {
            iNumberFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.NUMBER, aLocale);
        }
        return iNumberFormatKey;
    }
    /**
     * @return Returns the iTextFormatKey.
     */
    public int getTextFormatKey()
    {
        if (iTextFormatKey == -1)
        {
            iTextFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.TEXT, aLocale);
        }
        return iTextFormatKey;
    }
    /**
     * @return Returns the iTimeFormatKey.
     */
    public int getTimeFormatKey()
    {
        if (iTimeFormatKey == -1)
        {
            iTimeFormatKey = xNumberFormatTypes.getStandardFormat(NumberFormat.TIME, aLocale);
        }
        return iTimeFormatKey;
    }
}
